core/src/main/java/com/yahoo/oak/InternalOakMap.java

```diff
@@ -510,68 +527,73 @@ Result remove(K key, V oldValue, OakTransformer<V> transformer) {
+                        return ctx.result.withFlag(ValueUtils.ValueResult.FALSE);
+                    } else if (removeResult.operationResult == ValueUtils.ValueResult.RETRY) {
+                        continue;
+                    }
+                    // we have marked this value as deleted (successful remove)
+                    logicallyDeleted = true;
+                    v = (V) removeResult.value;
                 }
-                // we have marked this value as deleted (successful remove)
-                logicallyDeleted = true;
-                v = (V) removeResult.value;
-            }
 
-            // AT THIS POINT value was marked deleted off-heap by this thread,
-            // continue to set the entry's value reference as deleted
-            assert ctx.entryIndex != EntryArray.INVALID_ENTRY_INDEX;
-            assert ctx.isValueValid();
-            ctx.entryState = EntryArray.EntryState.DELETED_NOT_FINALIZED;
+                // AT THIS POINT value was marked deleted off-heap by this thread,
+                // continue to set the entry's value reference as deleted
+                assert ctx.entryIndex != EntryArray.INVALID_ENTRY_INDEX;
+                assert ctx.isValueValid();
+                ctx.entryState = EntryArray.EntryState.DELETED_NOT_FINALIZED;
 
-            if (inTheMiddleOfRebalance(c)) {
+                if (inTheMiddleOfRebalance(c)) {
+                    continue;
+                }
+
+                // If finalize deletion returns true, meaning rebalance was done and there was NO
+                // attempt to finalize deletion. There is going the help anyway, by next rebalance
+                // or updater. Thus it is OK not to restart, the linearization point of logical deletion
+                // is owned by this thread anyway and old value is kept in v.
+                finalizeDeletion(c, ctx); // includes publish/unpublish
+                return transformer == null ?
+                    ctx.result.withFlag(ValueUtils.ValueResult.TRUE) : ctx.result.withValue(v);
+            } catch (DeletedMemoryAccessException e) {
                 continue;
             }
-
-            // If finalize deletion returns true, meaning rebalance was done and there was NO
-            // attempt to finalize deletion. There is going the help anyway, by next rebalance
-            // or updater. Thus it is OK not to restart, the linearization point of logical deletion
-            // is owned by this thread anyway and old value is kept in v.
-            finalizeDeletion(c, ctx); // includes publish/unpublish
-            return transformer == null ?
-                ctx.result.withFlag(ValueUtils.ValueResult.TRUE) : ctx.result.withValue(v);
         }
 
         throw new RuntimeException("remove failed: reached retry limit (1024).");
     }
 
     // the zero-copy version of get
-    OakUnscopedBuffer get(K key) {
+    OakUnscopedBuffer get(K key) { //TODO what we do here?
```

Comment:
Here I would say to try/catch and to repeat again, but there is a chance to get into the infinite loop if nothing changes. In OakMap case we can get onto deleted key only if we are looking on the RELEASED chunk, and the released chunk shouldn't be found in skip-list anyway. So to ensure we are not gwtting to the same released chunk again and again we should accomplish the rebalance of this released chunk and to take it out of the skip-list, before next iteration of the loop.

Reply:
I mean we need to ensure method `inTheMiddleOfRebalance(BasicChunk<K, V> c)` is called within each loop iteration. This is also the answer to other `//TODO what we do here?` comments down this file